### PR TITLE
Add support for all the line feed control sequences

### LINKED ISSUE #3189: Support all the newline operations
# Description of the new feature/enhancement

VT terminals support at least five newline controls that I'm aware of. There are the three [control characters](https://vt100.net/docs/vt510-rm/chapter4.html#T4-1): `LF` (linefeed), `VT` (vertical tab), and `FF` (form feed). And then there are also two escape sequences: [`NEL` (Next Line)](https://vt100.net/docs/vt510-rm/NEL.html) and [`IND` (Index)](https://vt100.net/docs/vt510-rm/IND.html).

The three control characters are simply aliases for the same operation. When the [_New Line Mode_](https://vt100.net/docs/vt510-rm/LNM.html) is reset, they perform a linefeed by itself. When the _New Line Mode_ is set, they perform a carriage return as well as the linefeed.

The `NEL` and `IND` sequences are not affected by the mode, though. The `IND` escape sequence always performs a linefeed by itself, with no carriage return. And the `NEL` escape sequence always performs a carriage return and linefeed together.

It may not seem as if there's much value in supporting all of these sequences, but they are required to pass the [Vttest](https://invisible-island.net/vttest/) _Test of cursor movements_, so it would be nice to have them supported in the Windows console.

# Proposed technical implementation details (optional)

I'd ignore the _New Line Mode_ to start with, since that's not essential, and introduces a whole lot more complexity. And that means we could actually implement everything in the `OutputStateMachineEngine`, simply with calls to the `ITermDispatch::Execute` method, with either `LF`, or `CR` and `LF`, depending on what was needed.

That said, I think it would be nicer to add a dedicated `LineFeed` method to the `ITermDispatch` interface, probably with an enum parameter specifying the required type (i.e. with carriage return, without carriage return, or mode-dependent). Then we can always add a mode condition in there at a later stage if we want to support the [_New Line Mode_](https://vt100.net/docs/vt510-rm/LNM.html).

It  may even be worth adding a dedicated method in the `ConGetSet` interface for implementing the newline functionality directly, instead of going through `WriteCharsLegacy`. This requires a little more work, but has the advantage that it would not be influenced by the `DISABLE_NEWLINE_AUTO_RETURN` flag (which can produce the wrong behaviour when not set).

I should also mention there is an additional complication with implementing the `IND` sequence, which is issue #976. My preference would be to drop those cursor movement operations for now, in favor of getting `IND` working. And then we can always add them back when a proper VT52 mode is implemented (which is something I still have on the back burner). But it seems pointless keeping them without the rest of the VT52 support, and without the user explicitly requesting VT52 mode.

### PR DESCRIPTION
**Pull request by j4james**
## Summary of the Pull Request

This adds support for the `FF` (form feed) and `VT` (vertical tab) [control characters](https://vt100.net/docs/vt510-rm/chapter4.html#T4-1), as well as the [`NEL` (Next Line)](https://vt100.net/docs/vt510-rm/NEL.html) and [`IND` (Index)](https://vt100.net/docs/vt510-rm/IND.html) escape sequences.

## References

#976 discusses the conflict between VT100 Index sequence and the VT52 cursor back sequence.

## PR Checklist
* [x] Closes #3189
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [x] Tests added/passed
* [ ] Requires documentation to be updated
* [x] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #3189

## Detailed Description of the Pull Request / Additional comments

I've added a `LineFeed` method to the `ITermDispatch` interface, with an enum parameter specifying the required line feed type (i.e. with carriage return, without carriage return, or dependent on the [`LNM` mode](https://vt100.net/docs/vt510-rm/LNM.html)). The output state machine can then call that method to handle the various line feed control characters (parsed in the `ActionExecute` method), as well the `NEL` and `IND` escape sequences (parsed in the `ActionEscDispatch` method).

The `AdaptDispatch` implementation of `LineFeed` then forwards the call to a new `PrivateLineFeed` method in the `ConGetSet` interface, which simply takes a bool parameter specifying whether a carriage return is required or not. In the case of mode-dependent line feeds, the `AdaptDispatch` implementation determines whether the return is necessary or not, based on the existing _AutoReturnOnNewLine_ setting (which I'm obtaining via another new `PrivateGetLineFeedMode` method).

Ultimately we'll want to support changing the mode via the [`LNM` escape sequence](https://vt100.net/docs/vt510-rm/LNM.html), but there's no urgent need for that now. And using the existing _AutoReturnOnNewLine_ setting as a substitute for the mode gives us backwards compatible behaviour, since that will be true for the Windows shells (which expect a linefeed to also generate a carriage return), and false in a WSL bash shell (which won't want the carriage return by default).

As for the actual `PrivateLineFeed` implementation, that is just a simplified version of how the line feed would previously have been executed in the `WriteCharsLegacy` function. This includes setting the cursor to "On" (with `Cursor::SetIsOn`), potentially clearing the wrap property of the line being left (with `CharRow::SetWrapForced` false), and then setting the new position using `AdjustCursorPosition` with the _fKeepCursorVisible_ parameter set to false.

I'm unsure whether the `SetIsOn` call is really necessary, and I think the way the forced wrap is handled needs a rethink in general, but for now this should at least be compatible with the existing behaviour.

Finally, in order to make this all work in the _Windows Terminal_ app, I also had to add a basic implementation of the `ITermDispatch::LineFeed` method in the `TerminalDispatch` class. There is currently no need to support mode-specific line feeds here, so this simply forwards a `\n` or `\r\n` to the `Execute` method, which is ultimately handled by the `Terminal::_WriteBuffer` implementation.

## Validation Steps Performed

I've added output engine tests which confirm that the various control characters and escape sequences trigger the dispatch method correctly. Then I've added adapter tests which confirm the various dispatch options trigger the `PrivateLineFeed` API correctly. And finally I added some screen buffer tests that check the actual results of the `NEL` and `IND` sequences, which covers both forms of the `PrivateLineFeed` API (i.e. with and without a carriage return).

I've also run the _Test of cursor movements_ in the [Vttest](https://invisible-island.net/vttest/) utility, and confirmed that screens 1, 2, and 5 are now working correctly. The first two depend on `NEL` and `IND` being supported, and screen 5 requires the `VT` control character.

### CODE REVIEW

File: src/terminal/adapter/adaptDispatch.cpp [hunk 1]
```diff
@@ -1317,6 +1317,28 @@ bool AdaptDispatch::SetTopBottomScrollingMargins(const SHORT sTopMargin,
...
+// Arguments:
+// - lineFeedType - Specify whether a carriage return should be performed as well.
+// Return Value:
+// - True if handled successfully. False otherwise.
+bool AdaptDispatch::LineFeed(const DispatchTypes::LineFeedType lineFeedType)
+{
+    switch (lineFeedType)
+    {
+    case DispatchTypes::LineFeedType::DependsOnMode:
+        return !!_conApi->PrivateLineFeed(!!_conApi->PrivateGetLineFeedMode());
```
[the remark this answers is not in the document]
Comment (j4james) [reply src/terminal/adapter/adaptDispatch.cpp]: Yeah, I'm not convinced the current solution is perfect, but I did consider other approaches too. I tried passing the _lineFeedType_ straight through to the `ConGetSet::PrivateLineFeed` method, but it didn't seem like we could easily share the type across those interfaces. The only `ConGetSet` method that does something similar is `SetCursorStyle`, and that's achieved with two different enums and then having to map from the one to the other, which felt icky to me.

Using `std::optional<bool>` would avoid that problem - assuming we use that everywhere in place of the current enum - but it feels a little less readable. I'm not totally opposed to that approach though.

That said, another argument I had against handling the logic in `ConGetSet::PrivateLineFeed` was I wanted to try and keep that interface as simple as possible, and make `AdaptDispatch` handle anything VT-specific. The idea being that `ConGetSet` could one day be the interface that differentiates between the conhost and Terminal implementations, and all the generic VT behaviour would be shared in `AdaptDispatch` (what I was suggesting in issue #3849). Under those conditions you want as little code as possible in the `ConGetSet` implementation, because everything on that side needs to be written twice.

Bottom line is I have a slight preference for the current approach. But if you'd like me to go another way, I don't might changing things.

**Comment by j4james** [comment]
> I would sort of prefer losing the cursor move sequences in a separate commit before the new ones are introduced, but I don't think I'm going to lose that much sleep over it if it goes all in as one.

I've now cherry-picked the removal of the cursor movement ops into a separate PR (#4044), if you want to deal with that first.

File: src/terminal/adapter/adaptDispatch.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (miniksa) [reply src/terminal/adapter/adaptDispatch.cpp]: I'm willing to go with your existing slight preference.

File: src/host/getset.cpp [hunk 2]
```diff
@@ -1347,6 +1347,33 @@ void DoSrvPrivateAllowCursorBlinking(SCREEN_INFORMATION& screenInfo, const bool
...
+// - screenInfo - A pointer to the screen buffer that should perform the line feed.
+// - withReturn - Set to true if a carriage return should be performed as well.
+// Return value:
+// - STATUS_SUCCESS if handled successfully. Otherwise, an appropriate status code indicating the error.
+[[nodiscard]] NTSTATUS DoSrvPrivateLineFeed(SCREEN_INFORMATION& screenInfo, const bool withReturn)
+{
+    auto& textBuffer = screenInfo.GetTextBuffer();
+    auto cursorPosition = textBuffer.GetCursor().GetPosition();
+
+    textBuffer.GetCursor().SetIsOn(true);
```
Comment (DHowett-MSFT) [inline src/host/getset.cpp]: nit: this might be worth explicitly calling out with a comment; I'd hate it to get lost in a later refactor

Comment (j4james) [reply src/host/getset.cpp]: I'm not actually sure why it's necessary though, or even if it's really necessary. I was just copying the behavior from the previous `WriteCharsLegacy` codepath. But I'll do some more research and see if I can come up with something meaningful to say about it.

**Comment by j4james** [comment]
> Sorry though that merging #4005 created a ton of merge conflicts on this PR :/

No worries. I was expecting that. I'll try get it merged over the weekend.

File: src/host/getset.cpp (hunk 2 above)
Comment (j4james) [reply src/host/getset.cpp]: OK, it seems it's necessary to turn the cursor on to prevent old copies from sometimes being left behind when the screen scrolls. I've now added a comment explaining that.

**Comment by zadjii-msft** [comment]
@DHowett-MSFT Do we want to bring this in for 0.8? I'm inclined to hold it for 0.9 personally - just rather not have a change like this with as little runway as we have left. We might be able to tell the bot to merge this next week...

**Comment by DHowett-MSFT** [comment]
Let's hold on this one. I wonder if the bot can do that.

**Comment by DHowett-MSFT** [comment]
@msftbot merge this in 1 week

**Comment by DHowett-MSFT** [comment]
@msftbot merge this in 168 hours

**Comment by zadjii-msft** [comment]
@msftbot merge this in 1 minute

**Comment by DHowett-MSFT** [comment]
:tada: Once again, thanks for the contribution!

This pull request was included in a set of conhost changes that was just
released with [Windows Insider Build 19603](https://blogs.windows.com/windowsexperience/2020/04/08/announcing-windows-10-insider-preview-build-19603/).
